### PR TITLE
Stateless header chain verification

### PR DESCRIPTION
**Pull request by Zergity**
Lighting fast syncing and node performance.

### CODE REVIEW

File: consensus/dccs/1_snapshot.go [hunk 1]
```diff
@@ -51,12 +51,15 @@ func (s *Snapshot) init1() *Snapshot {
 
 // signers1 retrieves the list of authorized signers in hash ascending order.
 func (s *Snapshot) signers1() []Signer {
-	sigs := make([]Signer, 0, len(s.Signers))
-	for sig := range s.Signers {
-		sigs = append(sigs, Signer{Hash: s.Hash, Address: sig})
-	}
-	sort.Sort(signersAscendingByHash(sigs))
-	return sigs
+	s.sortedOnce.Do(func() {
```
Comment (hadv) [inline consensus/dccs/1_snapshot.go]: how can we re-sort when the signer list was changed?

File: consensus/dccs/1_dccs.go [hunk 2]
```diff
@@ -338,6 +417,47 @@ func (d *Dccs) prepare1(chain consensus.ChainReader, header *types.Header) error
 	return nil
 }
 
+// prepareBeneficiary gets the beneficiary of signer from smart contract and
+// set to header's coinbase to give sealing reward later.
+// if all else fails, the sealer address is kept as reward beneficiary
+func (d *Dccs) prepareBeneficiary(header *types.Header, chain consensus.ChainReader) {
+	index := common.BigToHash(common.Big1).String()[2:]
```
Comment (hadv) [inline consensus/dccs/1_dccs.go]: I intend to enforce the beneficiary in the next version. adding verify header

File: consensus/dccs/1_snapshot.go (hunk 1 above)
Comment (Zergity) [reply consensus/dccs/1_snapshot.go]: Signer list is re-sorted only on new checkpoint, which a new Snapshot object is created. Snapshot does not reused accrossed epoches.

File: consensus/dccs/1_dccs.go (hunk 2 above)
Comment (Zergity) [reply consensus/dccs/1_dccs.go]: I don't think enforcing coinbase address is a good idea, especially when I required the historical state (it's not cheap).

Comment (hadv) [reply consensus/dccs/1_dccs.go]: this new prepare has no much meaning and sealer might set beneficiary randomly without intend. As we already talked, enforce the beneficiary to ensure decentralised sealing pool. And in next version, we only need block header, no need state.

Comment (Zergity) [reply consensus/dccs/1_dccs.go]: We will discuss later the possibility of enforcing coinbase address for the next hardfork. This change is for the old consensus, which does not enforce it.

**Comment by Zergity** [comment]
Turn out, it's entitely possible, and fully secure, using the tx log receipts. 
The fastsync feature after the next hardfork is useless without fastsync support from all previous consensus versions.
This fastsync code is currently run on 2 of the testnet nodes, and a couple of the mainnet for awhile.

The tx receipts technique is currently used by my new consensus code, and latter will be applied to ThangLong consensus when I have the time.

**Comment by hadv** [comment]
> The tx receipts technique is currently used by my new consensus code, and latter will be applied to ThangLong consensus when I have the time.

okay, make sense but still not include in this PR. Luckily we fire event for join/leave sealer :)

**Pull request merged**